Sources/StreamChat/Repositories/SyncRepository.swift

```diff
@@ -0,0 +1,409 @@
+                    case let .failure(error):
+                        done(error.shouldRetry ? .retry : .continue)
+                    }
+                }
+            }
+        }
+        operations.append(contentsOf: refetchChannelOperations)
+
+        // 3. Refetch channel lists queries, link only what backend returns (the 1st page)
+        // 4. Clean up local message history for channels that are outdated/will get outdated
+        // We use `synchedChannelIds` to keep track of the channels that were synched both in the previous step and
+        // after each ChannelListController recovery.
+        let refetchChannelListQueryOperations: [AsyncOperation] = activeChannelListControllers.allObjects.compactMap { controller in
+            // Reset only the controllers that need recovery
+            guard controller.isAvailableOnRemote else { return nil }
+
+            return AsyncOperation(retries: retriesCount) { [weak self] done in
+                log.info("3. Refetch channel lists queries & 4. Clean up local message history", subsystems: .offlineSupport)
+                self?.resetChannelsQuery(
+                    for: controller,
+                    watchedChannelIds: watchedChannelIds,
+                    synchedChannelIds: synchedChannelIds
+                ) { result in
+                    switch result {
+                    case let .success(channels):
+                        let queryChannelIds = channels.map(\.cid)
+                        synchedChannelIds = synchedChannelIds.union(queryChannelIds)
+                        done(.continue)
+                    case let .failure(error):
+                        done(error.shouldRetry ? .retry : .continue)
+                    }
+                }
+            }
+        }
+        operations.append(contentsOf: refetchChannelListQueryOperations)
+
+        // 5. Bump the last sync timestamp
+        operations.append(AsyncOperation { [weak self] done in
+            log.info("5. Bump the last sync timestamp", subsystems: .offlineSupport)
+            self?.updateUserValue { user in
+                user?.lastSyncAt = Date()
+            } completion: { _ in
+                done(.continue)
+            }
+        })
+
+        operations.append(BlockOperation(block: {
+            log.info("Finished recovering offline state", subsystems: .offlineSupport)
+            DispatchQueue.main.async {
+                completion(nil)
+            }
+        }))
+
+        // We are making sure the operations happen secuentially one after the other by setting one as the dependency
+        // of the following one
+        var previousOperation: Operation?
+        operations.reversed().forEach { operation in
+            defer { previousOperation = operation }
+            guard let previousOperation = previousOperation else { return }
+            previousOperation.addDependency(operation)
+        }
+
+        operationQueue.addOperations(operations, waitUntilFinished: false)
```

Comment:
Question: what happens if the state quickly changes from connected -> disconnected -> connected again. Is there a possibility this method is called again while there are already running operations?

Reply:
Good catch! I'll investigate it now